### ISSUE TITLE
In-Game: Add support for "other" to be the originating "self" when used inside methods

### ISSUE THREAD

**Issue by katsaii** [YoYoGames/GameMaker-Bugs#7349]
### Description

For the following GML code:
```gml
global.a = { };
global.b = { };
global.c = { };

var f = method(c, function () {
  show_message(self == global.c);  // true
  show_message(other == global.b); // false,  expected true
  show_message(other == global.a); // true,   expected false
});

with (global.a) {
  with (global.b) {
    f();
  }
}
```

The `self` inside of the `f()` call is correctly `global.c`, but `other` ends up being `global.a`. After a discussion with Russell, this shouldn't be the case, and instead `other` should equal `global.b` in this example.

### Expected Change

Change `other` to be the "self" of the calling scope, when used inside of bound functions

### Steps To Reproduce

1. Start GameMaker
2. Create a new project
3. Create a new GML script
4. Copy-paste the example code into this script
5. Run the project
6. See the issue

### How reliably can you recreate this issue using your steps above?

Always

### Which version of GameMaker are you reporting this issue for?

2024.6.2 (Monthly)

### Which platform(s) are you seeing the problem on?

Windows, HTML5

### Contact Us Package Attached?

- [X] I have attached my Contact Us Package

### Sample Project Added?

- [ ] I have included a small sample project

**Comment by gnysek** [comment]
I think that it also doesn't work as it should in case of instances, and local functions, but I was waiting to check it and report correctly until #7125 comes into beta, as that could change a lot.

**Comment by rwkay** [comment]
This is not a bug as the `other` is not changed at the function call boundary the `other` is deliberately preserved at the method call for `f()` - it is the same `other` as at the call site (i.e. within the second `with`) - the `other` is not set to the `b` , this would be contrary to the that GML works at other script calls... i.e. within a collision.

**Comment by rwkay** [comment]
Use this project for testing...

[Bug-7349.zip](https://github.com/user-attachments/files/16804735/Bug-7349.zip)

**Comment by rwkay** [comment]
After discussion with Kat we have revised some of the rules around handling `other`.

The central principle is that when the current `self` is changed as part of the runtime then `other` will become the previous `self`, until now this has only happened in 2 places 

1. As part of the `with` statement
2. When `new` is called when the constructor is executed `other` is set to the `self` at the point that `new` was called.

To fit with the central principle we have introduced another point where `other` changes and expanded on 2 above. So the revised rules mean that `other` now changes when.

1. As part of the `with` statement
2. When a function is called as part of a bound `method` - `other` is set to the original `self` at the call site.
3. When `new` is called if the constructor is just a function then as before `other` is set to the `self` at the call site, if the constructor is a bound `method` then `other` is the bound `self`.

We will make this change in 2024.10 - we expect the impact will be minimal (let us know if that is not the case)

**Comment by rwkay** [comment]
Added in 2024.10

**Comment by KormexGit** [comment]
That new case sounds super useful!

**Comment by mgeddesGM** [comment]
verified as of  IDE v2024.1100.0.626  Runtime v2024.1100.0.652

**Comment by KormexGit** [comment]
This appears to only be implemented in YYC, not VM, in 2024.1100.0.652.

Testing with the following code:
```
function test() constructor {
	draw = function() {
		show_debug_message(other)
	}
}

//on an object, create event
struct = new test();
//draw event
struct.draw();
```
In YYC, the message prints the struct:
```
{ draw : function }
```
In VM, it prints the object:
```
{ struct : { draw : function gml_Script_anon@39@test@Script1 } }
```

[Only YYC Bug.zip](https://github.com/user-attachments/files/17124745/Only.YYC.Bug.zip)

**Comment by rwkay** [comment]
Updated fix in 2024.11

**Comment by scott-dunbar** [comment]
Verified across vm,yyc,html5 on all attached projects

 IDE v2024.1100.0.646  Runtime v2024.1100.0.666
